File: week4_paths2/2_negative_cycle/negative_cycle.py
#Uses python3

import sys


class Directed_Graph:
    def __init__(self):
        self.adj = None
        self.cost = None

    def Data_Read(self):
        input = sys.stdin.read()
        data = list(map(int, input.split()))
        n, m = data[0:2]
        data = data[2:]
        edges = list(zip(zip(data[0:(3 * m):3], data[1:(3 * m):3]), data[2:(3 * m):3]))
        data = data[3 * m:]
        self.adj = [[] for _ in range(n+1)]
        self.cost = [[] for _ in range(n+1)]
        for ((a, b), w) in edges:
            self.adj[a - 1].append(b - 1)
            self.cost[a - 1].append(w)

        '''
            What if the negative cycle is unreachable from the source vertex?
            The simplest approach is to add an extra vertex and add an edge of length 1 from this vertex to each vertex of the initial graph.
            you have to set the source to be the added vertex.
        '''
        for i in range(n):
            self.adj[n].append(i)
            self.cost[n].append(1)

    def is_negative_cycle(self):
        ''' Detect the negative_cycle in the given connected graph.
            Time Complexity: O(|V||E|)
        '''
        dist = [float('inf')] * len(self.adj)
        source = len(self.adj) - 1
        dist[source] = 0

        for k in range(len(self.adj)):  # |V| iterations
            for i in range(len(self.adj)):
                for j in range(len(self.adj[i])):
                    u = i
                    v = self.adj[i][j]
                    w = self.cost[i][j]  # from u to v

                    # check for relaxation
                    # Note that only if u is already relaxed, v can be relaxed.
                    if dist[u] != float('inf') and dist[v] > dist[u] + w:
                        if k == len(self.adj) - 1:
                            return 1

                        dist[v] = dist[u] + w;

        return 0




def negative_cycle(adj, cost):
    #write your code here
    return 0


if __name__ == '__main__':
    graph = Directed_Graph()  # () is necessary but not for C++
    graph.Data_Read()
    print(graph.is_negative_cycle())
    # input = sys.stdin.read()
    # data = list(map(int, input.split()))
    # n, m = data[0:2]
    # data = data[2:]
    # edges = list(zip(zip(data[0:(3 * m):3], data[1:(3 * m):3]), data[2:(3 * m):3]))
    # data = data[3 * m:]
    # adj = [[] for _ in range(n)]
    # cost = [[] for _ in range(n)]
    # for ((a, b), w) in edges:
    #     adj[a - 1].append(b - 1)
    #     cost[a - 1].append(w)
    # print(negative_cycle(adj, cost))

File: week4_paths2/2_negative_cycle/negative_cycle.cpp
#include <iostream>
#include <vector>
#include <limits>
using namespace std;
const int int_max = numeric_limits<int>::max();

using std::vector;
class Directed_Graph
{
  private:
    vector<vector<int>> adj;
    vector<vector<int>> cost;
  
  public:
    void Data_Read()
    {
      int n, m;
      std::cin >> n >> m;
      adj.resize(n);
      cost.resize(n);
      /* vector<vector<int> > adj(n, vector<int>()); */
      /* vector<vector<int> > cost(n, vector<int>()); */
      for (int i = 0; i < m; i++) {
	int x, y, w;
	std::cin >> x >> y >> w;
	adj[x - 1].push_back(y - 1);
	cost[x - 1].push_back(w);
      }

    }
    void ensure_connected(vector<vector<int>> &adj_augmented, vector<vector<int>>  &cost_augmented)
    {
      /*
	 What if the negative cycle is unreachable from the source vertex?
	 The simplest approach is to add an extra vertex and add an edge of length 1 from this vertex to each vertex of the initial graph.
	 you have to set the source to be the added vertex.
      */
      adj_augmented.resize(adj.size() + 1);
      cost_augmented.resize(adj.size() + 1);
      int n = adj.size();
      for(int i(0); i < adj.size(); i++)
	for(int j(0); j < adj[i].size(); j++)
	{
	  adj_augmented[i].push_back(adj[i][j]);
	  cost_augmented[i].push_back(cost[i][j]);
	}

      for (int i = 0; i < n; i++)
      {
	adj_augmented[n].push_back(i);
	cost_augmented[n].push_back(1);
      }
    }

    int is_negative_cycle()
    {
      /*
	 Only holds for connected graph.
      */
      vector<vector<int>> adj_augmented;
      vector<vector<int>> cost_augmented;
      ensure_connected(adj_augmented, cost_augmented);
      vector<int> dist(adj_augmented.size(), int_max);
      int source = adj_augmented.size() - 1; // Starting from the added vertex, you can reach the originally unreachable negative cycle.
      dist[source] = 0;
      for(int k(1); k <= adj_augmented.size(); k++)
      {
	for(int i(0); i < adj_augmented.size(); i++)
	  for(int j(0); j < adj_augmented[i].size(); j++)
	  {
	    int u = i;
	    int v = adj_augmented[i][j];
	    int w = cost_augmented[i][j]; //  from u to v

	    // relax
	    if(dist[v] > dist[u] + w && (dist[u] != int_max))
	    {
	      if(k == adj_augmented.size())
	      {
		return 1;
	      }
	      dist[v] = dist[u] + w;
	    }

	  }
      }
      return 0;
    }

};

int main() {
  Directed_Graph graph;
  graph.Data_Read();
  cout << graph.is_negative_cycle() << endl;
  /* int n, m; */
  /* std::cin >> n >> m; */
  /* vector<vector<int> > adj(n, vector<int>()); */
  /* vector<vector<int> > cost(n, vector<int>()); */
  /* for (int i = 0; i < m; i++) { */
  /*   int x, y, w; */
  /*   std::cin >> x >> y >> w; */
  /*   adj[x - 1].push_back(y - 1); */
  /*   cost[x - 1].push_back(w); */
  /* } */
  /* std::cout << negative_cycle(adj, cost); */
}
